src/lib/util/graceful-shutdown.ts

```diff
@@ -1,7 +1,60 @@
-import { Logger } from '../logger';
+import { Logger, LogProvider } from '../logger';
 import { IUnleash } from '../types/core';
+import { GenericUnleashError } from '../error/unleash-error';
 
-function registerGracefulShutdown(unleash: IUnleash, logger: Logger): void {
+export interface IGracefulShutdown {
+    registerGracefulShutdownHook(
+        serviceName: string,
+        hook: () => Promise<void>,
+    ): void;
+    resetShutdownHooks(): void;
+    executeShutdownHooks(): Promise<void>;
+}
+
+export class GracefulShutdownHookManager implements IGracefulShutdown {
+    private shutdownHooks: Map<string, () => Promise<void>>;
+    private logger: Logger;
+    constructor(getLogger: LogProvider) {
+        this.shutdownHooks = new Map<string, () => Promise<void>>();
+        this.logger = getLogger('/lib/util/graceful-shutdown.ts');
+    }
+
+    async executeShutdownHooks(): Promise<void> {
+        for (const [hookName, hook] of this.shutdownHooks.entries()) {
+            this.logger.info(`Shutting down ${hookName}`);
+            try {
+                await hook();
+                this.shutdownHooks.delete(hookName);
+                this.logger.info(`Done shutting down ${hookName}`);
+            } catch (e) {
+                this.logger.error(`Shutdown hook ${hookName} failed`, e);
+            }
+        }
+    }
+
+    registerGracefulShutdownHook(
+        serviceName: string,
+        hook: () => Promise<void>,
+    ): void {
+        this.logger.info(`Registering graceful shutdown for ${serviceName}`);
+        if (this.shutdownHooks.has(serviceName)) {
+            throw new GenericUnleashError({
+                name: 'DuplicateShutdownHook',
+                message: `${serviceName} already registered for graceful shutdown`,
+                statusCode: 500,
+            });
```

Comment:
Might be cleaner to create a new `src/lib/error/duplicate-shutdown-hook-error.ts`

Reply:
Yeah, fixed in cb54915dd